---
title: "dj1KO_reanalysis"
author: "Karissa"
date: "24/02/2021"
output: 
  html_document: 
    fig_caption: yes
    fig_height: 6
    fig_width: 8
    toc: yes
editor_options: 
  chunk_output_type: console

---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE, warning = FALSE, message = FALSE)
```

## packages

```{r}
# analysis and wrangling 
library(tidyverse)
library(magrittr)
library(edgeR)
library(readxl)
library(AnnotationHub)
library(limma)

# data vis
library(RColorBrewer)
library(pheatmap)
library(kableExtra)
library(pander)

# ggplot extensions
library(ggrepel)
library(ggeasy)
library(ggfortify) 
library(ggpubr)
```

# Introduction

```{r, fig.cap="Effect of filterling lowly expressed genes"}
counts <-  list.files("salmonOut/", full.names = TRUE) %>%
  catchSalmon() %>%
  .$counts %>%
  as.data.frame() %>%
  rownames_to_column("transcript_id") %>%
  as_tibble()

anno <-read_tsv("ncbiRefSeq.txt", col_names = FALSE)

dge <- anno %>%
  .[,c(2,13)] %>%
  set_colnames(c("transcript_id", "gene_name")) %>%
  left_join(counts) %>%
  na.omit %>%
  gather(key = "Sample", value = "Counts", starts_with("salmon")) %>%
  mutate(Sample = str_remove(Sample, pattern = 'salmonOut//')) %>%
  group_by(Sample, gene_name) %>%
  summarise(Counts = sum(Counts)) %>%
  spread(key = "Sample", value = "Counts") %>%
  column_to_rownames("gene_name") %>%
  DGEList() %>%
  calcNormFactors(method = "TMM")

dge$samples %<>%
  mutate(sample = rownames(.),
         Genotype = factor(x = c(rep("WT", 3), rep("dj1", 3)), levels = c("WT", "dj1"))
  )
```

The aim of this analysis is to assess whether the dj1-/- zebrafish brains show evidence for iron dyshomeostasis. To test for this, I will take the processed data from Hughes et al., and perform enrichment analysis on the IRE gene sets from Nhi Hin to test whether the IRE genes show significant differential expression as a group. 

First, I will read in the data from Hughes et al. using the `catchSamlon` function from the `edgeR` package. I will omit lowly expressed genes from the analysis. Here, I will omit genes which contain at least 1 CPM in at least 3 out of the 6 samples. This equated to about `r floor(mean(1*dge$samples$lib.size/1e6))`

Hughes et al, used the NCBI annotations. so I obtained the annoatation file from the NCBI website. 

The effect of filtering lowly expressed genes is shown in the plots below. 


```{r, fig.cap="Effect of filterling lowly expressed genes"}
a <- 
  dge %>% 
   cpm(log = TRUE) %>%
  as.data.frame() %>%
  rownames_to_column('gene_name') %>%
  gather(key = "sample", value = 'logCPM', 2:7) %>% 
  left_join(dge$samples) %>%
  ggplot(aes(x = logCPM, colour = sample)) +
  geom_density() +
  guides(colour = FALSE) +
  ggtitle("Before filtering") +
  labs(x = "logCPM", y = "Proportion of Genes") +
  theme_bw()

# filter lowly expressed genes
keepThesegenes <- (rowSums(cpm(dge) > 1) >= 3)
dge <- dge[keepThesegenes,, keep.lib.sizes=FALSE]


ggarrange(a, 
          dge %>%
            cpm(log = TRUE) %>%
            as.data.frame() %>%
            rownames_to_column('gene_name') %>%
            gather(key = "sample", value = 'logCPM', 2:7) %>%
            left_join(dge$samples) %>%
            ggplot(aes(x = logCPM, colour = sample)) +
            geom_density() +
            guides(colour = FALSE) +
            ggtitle("After filtering") +
            labs(x = "logCPM", y = "Proportion of Genes") +
            theme_bw()
)
```
# Principal component analysis

Next, I will assess the overall similarity between samples. Sample `GH2`, a wild type sample, appears very distant from the rest and may be an outlier. However, since n = 3, i will retain it in the analysis. 

```{r}
cpm(dge, log = TRUE) %>%
  t() %>%
  prcomp %>%
  autoplot(data = tibble(sample = rownames(.$x)) %>%
             left_join(dge$samples),
           colour = "Genotype",
           size = 4
  ) +
  geom_text_repel(
    aes(label = sample, colour = Genotype),
    show.legend = FALSE
  ) +
  theme_bw()
```

# DGE

Here, I will perform a DE analysis using `edgeR`. 
```{r}
design <- model.matrix(~Genotype, data = dge$samples)

fit <- dge %>%
  estimateDisp(design) %>%
  glmFit(design)


toptable <- glmLRT(fit, coef = "Genotypedj1") %>%
  topTags(n = Inf) %>%
  .[["table"]] %>%
  as.data.frame() %>%
  rownames_to_column('gene_name') %>%
  as_tibble() %>%
  arrange(PValue) %>%
  mutate(
    DE = FDR < 0.05
  )
```

```{r, fig.cap="Volcano plot of DE genes in dj1-/- brains"}
toptable %>%
  ggplot(aes(x = logFC, y = -log10(PValue), colour = DE)) +
  geom_point(alpha =0.5) +
  theme_bw() +
  scale_y_continuous(limits = c(0, 10)) +
  geom_label_repel(
    aes(label = gene_name),
    data = .  %>% dplyr::filter(FDR < 0.05),
    show.legend = FALSE
  ) + 
  scale_color_manual(values = c("grey50", "red"))

```

# IRE test using `fry`

The IRE gene sets were obtained from Nhi Hin (see her IRE paper on biorxiv). The RDS file i have is in ensembl gene id format. So I will convert the ensID to gene name using `AnnotationHub`. Then, IRE genes were restricted to only the genes tested in the DE analysis (i.e. detectable in the RNA-seq experiment)

```{r}
# import ire gene sets
zebrafishIreGenes <- readRDS("zebrafishIreGenes.rds")

# convert this R onject rom ensID to gene_name using annotationHub
ah <- AnnotationHub() %>%
  subset(species == "Danio rerio") %>%
  subset(rdataclass == "EnsDb")
ensDb <- ah[["AH74989"]]
grGenes <- genes(ensDb)

ire <- zebrafishIreGenes %>%
  lapply(function(x){
    x %>%
      as_tibble() %>%
      set_colnames("gene_id") %>%
      left_join(grGenes %>%
                  as_tibble()
                ) %>%
      dplyr::filter(symbol %in% rownames(dge$counts)) %>%
      .$symbol
  })
```


```{r}
fry(cpm(dge, log = TRUE), index = ire, design = design, contrast = "Genotypedj1") %>%
  as.data.frame() %>%
  rownames_to_column("IRE") %>%
  ggplot(aes(x = -log10(PValue), y = IRE, fill = FDR < 0.05)) +
  geom_col() +
  geom_vline(xintercept = -log10(0.03)) +
  scale_fill_viridis_d(end = 0.6) +
  theme_bw() 
  #ggsave("pvalue.png", width = 20, height = 10, units = "cm", dpi = 800, scale = 0.5)



# png("heatmap.png", width = 30, height = 5, units = "cm", res = 800)
toptable %>%
  dplyr::filter(gene_name %in% ire$ire5_hq) %>%
  dplyr::select(gene_name, logFC) %>%
  column_to_rownames("gene_name") %>%
  arrange(desc(logFC)) %>% 
  t() %>% 
  pheatmap(cluster_cols = F, cluster_rows = F,
           main = "IRE 5' HQ",
           cellwidth = 15, cellheight = 15, angle_col = 45,
           color = colorRampPalette(rev(brewer.pal(n = 7, name = 'RdBu')))(100)
  )
```
# Session info
```{r}
sessionInfo() %>% pander()
```





